ratatool-sampling/src/main/scala/com/spotify/ratatool/samplers/util/SamplerSCollectionFunctions.scala

```diff
@@ -0,0 +1,280 @@
+      .filter{case (_, ((_, d), t)) => d <= t}
+      .map{case (k, ((v, _), _)) => (k, v)}
+  }
+
+  private def stratifiedThresholdByKey[T: ClassTag, U: ClassTag](
+      s: SCollection[(U, (T, Double))],
+      countsByKey: SCollection[(U, (Long, Long))],
+      targetByKey: SCollection[(U, Long)],
+      variance: Double,
+      prob: Double)
+  : SCollection[(U, Double)] = {
+    s.map{case (k, (_, d)) => (k, d)}
+      .filter{case (_, d) => d < (prob + variance) && d > (prob - variance)}
+      //TODO: Clean up magic number
+      .topByKey(1e8.toInt)(Ordering.by(identity[Double]).reverse)
+      .hashJoin(countsByKey)
+      .hashJoin(targetByKey)
+      .map { case (k, ((itr, (lower, upper)), target)) =>
+        if (lower >= target) {
+          (k, prob - variance)
+        }
+        else if (upper < target) {
+          (k, prob + variance)
+        }
+        else {
+          (k, itr.drop((target - lower).toInt).headOption.getOrElse(prob + variance))
+        }
+    }
+  }
+
+  private def uniformThresholdByKey[T: ClassTag, U: ClassTag](s: SCollection[(U, (T, Double))],
+                                                        countsByKey: SCollection[(U, (Long, Long))],
+                                                        varByKey: SCollection[(U, Double)],
+                                                        probByKey: SCollection[(U, Double)],
+                                                        popPerKey: SideInput[Double])
+  : SCollection[(U, Double)] = {
+    s.map{ case (k, (_, d)) => (k, d) }
+      .hashJoin(varByKey)
+      .hashJoin(probByKey)
+      .filter{case (_, ((d, variance), prob)) =>
+        d < (prob + variance) && d >= (prob - variance)}
+      .map{case (k, ((d, _), _)) => (k, d)}
+      //TODO: Clean up magic number
+      .topByKey(1e8.toInt)(Ordering.by(identity[Double]).reverse)
+      .hashJoin(countsByKey)
+      .hashJoin(varByKey)
+      .hashJoin(probByKey)
+      .withSideInputs(popPerKey)
+      .map { case ((k, ((((itr, (lower, upper)), variance)), prob)), sic) =>
+        if (lower >= sic(popPerKey)) {
+          (k, prob - variance)
+        }
+        else if (upper < sic(popPerKey)) {
+          (k, prob + variance)
+        }
+        else {
+          val threshold = itr.drop(max(0, (sic(popPerKey) - lower).toInt)).headOption
+          (k, threshold.getOrElse(prob + variance))
+        }
+      }.toSCollection
+  }
+
+  implicit class RatatoolKVDSCollection[T: ClassTag, U: ClassTag](s: SCollection[(U, (T, Double))]){
+    //scalastyle:off cyclomatic.complexity
+    /**
+     * Performs higher precision sampling for a given distribution by assigning random probabilities
+     * to each element in SCollection[T] and then finding the Kth lowest, where K is the target
+     * population for a given strata or key.
+     *
+     * Currently variance is used to create cut-offs for upper and lower bounds to find K,
+     * but this can be improved inthe future.
```

Comment:
could you clarify what's suboptimal / needs improvement about this

Reply:
No longer relevant since we are now using better bounds, but basically the wider the bounds, the more "coin flips" we need to fit into memory if that makes sense